File: cpp/src/benchmarks/threads.cpp
#include "threads.h"
#include <atomic>
#include <cassert>
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

namespace bch {
void benchmark_threads()
{
    constexpr unsigned long long n_threads = 10;
    constexpr unsigned long long n_iter    = 1'000'000;
    unsigned long long           counter   = 0;
    std::vector<std::thread>     threads;
    threads.reserve(n_threads);
    std::mutex mx;
    for (std::size_t i = 0; i != n_threads; ++i) {
        threads.emplace_back([n_iter, &counter, &mx]() {
            for (std::size_t i = 0; i != n_iter; ++i) {
                std::lock_guard<std::mutex> guard{mx};
                ++counter;
            }
        });
    }
    for (std::size_t i = 0; i != n_threads; ++i) {
        threads[i].join();
    }
    assert(counter == n_threads * n_iter);
}
} // namespace bch

File: rust/src/benchmarks/threads.rs
use std::sync::{Mutex, Arc};
use std::thread;

pub fn benchmark_threads() -> ()
{
    const N_THREADS: u64 = 10;
    const N_ITER: u64 = 1_000_000;
    let counter = Arc::new(Mutex::new(0 as u64));
    let mut threads = Vec::new();
    for _i in 0..N_THREADS {
        let counter = counter.clone();
        threads.push(thread::spawn(move || {
            for _j in 0..N_ITER {
                let mut data = counter.lock().unwrap();
                *data += 1;
            }
        }));
    }
    for thread in threads {
        thread.join().unwrap();
    }
    assert!(*counter.lock().unwrap() == N_THREADS*N_ITER);
}
